File: go/rate.go
package walletpay

import (
	"context"

	"github.com/openweb3-io/wallet-pay-openapi/go/internal/openapi"
)

type (
	EstimateIn   = openapi.EstimateData
	EstimateOut  = openapi.EstimateOut
	RatesOut     = openapi.RatesOut
	GetRatesIn   = openapi.GetRatesIn
	CurrencyPair = openapi.CurrencyPair
)

type Rate struct {
	api *openapi.APIClient
}

type EstimateOptions struct {
	BaseCurrency string
	BaseAmount   string
	ToCurrency   string
}

func (e *Rate) Estimate(ctx context.Context, options *EstimateOptions) (*EstimateOut, error) {
	req := e.api.RateApi.V1RateEstimate(ctx)
	req = req.BaseCurrency(options.BaseCurrency)
	req = req.BaseAmount(options.BaseAmount)
	req = req.ToCurrency(options.ToCurrency)
	out, res, err := req.Execute()
	if err != nil {
		return nil, wrapError(err, res)
	}
	ret := EstimateOut(out.Data)
	return &ret, nil
}

func (e *Rate) GetRates(ctx context.Context, getRatesIn *GetRatesIn) (*RatesOut, error) {
	req := e.api.RateApi.V1RateGetRates(ctx)
	req = req.GetRatesIn(openapi.GetRatesIn(*getRatesIn))
	out, res, err := req.Execute()
	if err != nil {
		return nil, wrapError(err, res)
	}
	ret := RatesOut(out.Data)

	return &ret, nil
}

File: java/lib/src/main/java/io/openweb3/walletpay/Rate.java
package io.openweb3.walletpay;

import io.openweb3.walletpay.exceptions.ApiException;
import io.openweb3.walletpay.internal.api.RateApi;
import io.openweb3.walletpay.models.RatesOut;
import io.openweb3.walletpay.models.EstimateOut;
import io.openweb3.walletpay.models.GetRatesIn;

public final class Rate {
	private final RateApi api;

	Rate() {
		api = new RateApi();
	}

	public EstimateOut estimate(final RatesOptions options) throws ApiException {
		try {
			return api.v1RateEstimate(options.getBaseCurrency(), options.getToCurrency(), options.getBaseAmount()).getData();
		} catch (io.openweb3.walletpay.internal.ApiException e) {
			throw Utils.WrapInternalApiException(e);
		}
	}

	public RatesOut getRates(final GetRatesIn ratesIn) throws ApiException {
		try {
			return api.v1RateGetRates(ratesIn).getData();
		} catch (io.openweb3.walletpay.internal.ApiException e) {
			throw Utils.WrapInternalApiException(e);
		}
	}
}
